src/Html/HtmlCssSelectors.fs

```diff
@@ -0,0 +1,375 @@
+﻿namespace FSharp.Data
+
+open System
+open FSharp.Data
+open System.Runtime.CompilerServices
+
+module internal HtmlCssSelectors =
+    type SelectorToken =
+        | ClassPrefix of int
+        | IdPrefix of int
+        | TagName of int * string
+        | CssClass of int * string
+        | CssId of int * string
+        | AllChildren of int
+        | OpenAttribute of int 
+        | CloseAttribute of int
+        | AttributeName of int * string
+        | AttributeValue of int * string
+        | Assign of int
+        | EndWith of int
+        | StartWith of int
+        | DirectChildren of int 
+        | AttributeContainsPrefix of int
+        | AttributeContains of int
+        | AttributeContainsWord of int
+        | AttributeNotEqual of int
+        | Checkbox of int
+        | Checked of int
+        | Disabled of int
+        | Enabled of int
+        | Selected of int
+        | Button of int
+        | EmptyNode of int
+        | File of int
+        | Hidden of int
+        | Radio of int
+        | Password of int
+
+    type CssSelectorTokenizer() =
+        let mutable charCount:int = 0
+        let mutable source = List<char>.Empty
+        let mutable cssSelector = ""
+        let mutable inQuotes:bool = false
+
+        let getOffset (t:List<char>) = 
+                charCount - 1 - t.Length
+
+        member public x.Tokenize(pCssSelector:string) =
+            cssSelector <- pCssSelector
+            source <- Array.toList(cssSelector.ToCharArray())
+            charCount <- source.Length
+            x.tokenize()
+            
+        member private x.tokenize() = 
```

Comment:
For parsing speed, I tend to use a struct like
```
[<Struct>]
type State(source: string, pos: string) =
    member __.Source = source
    member __.Pos = pos
```
and a module with some inline functions, like next etc.. And an inline active record to match current char and tail..
You don't get the :: syntax so it leads to more parenthesis nesting.. but it works quite well.

Since css selector are usually short, I  don't know if it'll make much difference anyway..